bluemira/optimisation/_nlopt/optimiser.py

```diff
@@ -281,6 +270,33 @@ def set_upper_bounds(self, bounds: Union[np.ndarray, float]) -> None:
         for constraint in self._eq_constraints + self._ineq_constraints:
             constraint.set_approx_derivative_upper_bound(bounds)
 
+    def _handle_round_off_error(self) -> Tuple[np.ndarray, float]:
+        """
+        Handle a round-off error occurring in an optimisation.
+
+        It's likely the last call was a decent solution, so return that
+        (with a warning).
+        """
+        round_off_msg = "optimisation: round-off error occurred"
+        if self._objective.history:
+            # TODO(hsaunders1904): Is this really what we want to
+            # return? Can't the minimum have been something that
+            # wildly violates constraints?
```

Comment:
It's possible, I think, depending on the algorithm. But normally round-off errors return reasonable points.

Quoting the NLOpt docs here https://nlopt.readthedocs.io/en/latest/NLopt_Reference/#error-codes-negative-return-values:
```
NLOPT_ROUNDOFF_LIMITED = -4

Halted because roundoff errors limited progress. (In this case, the optimization still typically returns a useful result.)

```

Reply:
Sorry, that was not a good answer. Upon round-off, we return best point found so far, rather than the last. This is what we were doing before, correct? I guess I got paranoid... In my experience, round-off errors tend to occur if:
* you haven't normalised the input vector and have either very large or very small (or a mixture of both)
* you are more or less stepping around an optimum under certain conditions (e.g. near one or more constraints).
Perhaps it would be wiser to follow whatever NLopt does here, or check the best vector against the constraints.

Reply:
OK, that makes sense. Is what we're doing okay here? 

The old optimiser stuff was just returning a reference to the parameterisation array, not the best result so far. I'm not convinced it was returning the parameterisation from the previous iteration, but rather returning the parameterisation that caused the round-off error (but maybe that's OK?). I'll need to double-check. At the moment, we're copying the parametrisation at every iteration, which is not ideal.

Busy with STEP stuff today and tomorrow, so probably won't get to this until next week.

Reply:
It was found that the old optimiser was returning uninitialised memory on a round-off error.

It seems that we cannot get around making a copy of the parameterisation at every iteration if we want to return the last result on a round-off error. So that's what we do in the new optimiser.

<details>
  <summary>Round-off error snippet (thanks @CoronelBuendia)</summary>

  ```python
    class RoundOffErr:
      def __init__(self) -> None:
          self.it = 0
  
      def __call__(self, x, grad):
          """Objective function to create a round-off error."""
          self.it += 1
          value = -x[0] + x[1]
  
          if grad.size > 0:
              grad[0] = -1
              grad[1] = 1
  
          print(f"{self.it:02}: {x=}, {value=}")
          return value
  

  f_round_off = RoundOffErr()
  optimiser = Optimiser("COBYLA", 2, opt_conditions={"ftol_abs": 0.01})
  optimiser.set_objective_function(f_round_off)
  optimiser.set_lower_bounds([-3, -3])
  optimiser.set_upper_bounds([3, 3])
  result = optimiser.optimise([0.5, 2])
  print(f"{result=}")
  print(f"{optimiser.optimum_value=}")
  ```
</details>